src/database/mock.rs

```diff
@@ -92,76 +92,83 @@ impl MockDatabase {
+        match self.query_results.len() {
+            len if counter < len => Ok(std::mem::take(&mut self.query_results[counter])
                 .into_iter()
                 .map(|row| QueryResult {
                     row: QueryResultRow::Mock(row),
                 })
-                .collect())
-        } else {
-            Err(DbErr::Query(RuntimeErr::Internal(
+                .collect()),
+            _ => Err(DbErr::Query(RuntimeErr::Internal(
                 "`query_results` buffer is empty.".to_owned(),
-            )))
+            ))),
         }
     }
 
     #[instrument(level = "trace")]
-    fn begin(&mut self) {
-        if self.transaction.is_some() {
-            self.transaction
-                .as_mut()
-                .unwrap()
-                .begin_nested(self.db_backend);
-        } else {
-            self.transaction = Some(OpenTransaction::init());
-        }
+    fn begin(&mut self) -> Result<(), DbErr> {
+        match self.transaction.as_mut() {
+            Some(transaction) => transaction.begin_nested(self.db_backend),
+            None => self.transaction = Some(OpenTransaction::init()),
+        };
+        Ok(())
     }
 
     #[instrument(level = "trace")]
-    fn commit(&mut self) {
-        if self.transaction.is_some() {
-            if self.transaction.as_mut().unwrap().commit(self.db_backend) {
-                let transaction = self.transaction.take().unwrap();
-                self.transaction_log.push(transaction.into_transaction());
+    fn commit(&mut self) -> Result<(), DbErr> {
+        match self.transaction.as_mut() {
+            Some(transaction) => {
+                if transaction.commit(self.db_backend) {
+                    let transaction = self
+                        .transaction
+                        .take()
+                        .expect("No open transaction to commit");
+                    self.transaction_log.push(transaction.into_transaction()?);
+                }
+                Ok(())
             }
-        } else {
-            panic!("There is no open transaction to commit");
+            None => Err(DbErr::Exec(RuntimeErr::Internal(
+                "There is no open transaction to commit".to_owned(),
+            ))),
         }
     }
 
     #[instrument(level = "trace")]
-    fn rollback(&mut self) {
-        if self.transaction.is_some() {
-            if self.transaction.as_mut().unwrap().rollback(self.db_backend) {
-                let transaction = self.transaction.take().unwrap();
-                self.transaction_log.push(transaction.into_transaction());
+    fn rollback(&mut self) -> Result<(), DbErr> {
+        match self.transaction.as_mut() {
+            Some(transaction) => {
+                if transaction.rollback(self.db_backend) {
+                    let transaction = self
+                        .transaction
+                        .take()
+                        .expect("No open transaction to rollback");
```

Comment:
Maybe is better return `Err` instead `panic`?

Reply:
Nice catch!

Reply:
@billy1624 why did you close conversation?

Reply:
Oh I for got to push